### PR TITLE
[WIP] Add context to all relevant cloud APIs.

### LINKED ISSUE #815: Proposal: Use context throughout kubernetes
In an ideal world we need a way to seamlessly carry trace, debugging, cancellation and other data across boundaries - whether that be process, goroutine or RPC.

The code.google.com/p/go.net/context package provides wiring to do this.
References http://blog.golang.org/context

I'd like to wire this into kubernetes, but want blessing before going off and doing it. If you think this would be of value let me know and I'll send a PR.

### PR DESCRIPTION
**Pull request by cheftako**
**What this PR does / why we need it**: This adds context to all the relevant cloud provider interface signatures. It also propagates that change to all calles

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #815

**Special notes for your reviewer**:

**Release note**:
```
All implementer of the cloud provider interface will have to fix their implementation to account for the Context.
```

### CODE REVIEW

File: pkg/cloudprovider/providers/aws/aws.go [hunk 1]
```diff
@@ -1440,7 +1441,7 @@ func (c *Cloud) GetZone() (cloudprovider.Zone, error) {
 // GetZoneByProviderID implements Zones.GetZoneByProviderID
 // This is particularly useful in external cloud providers where the kubelet
 // does not initialize node data.
-func (c *Cloud) GetZoneByProviderID(providerID string) (cloudprovider.Zone, error) {
+func (c *Cloud) GetZoneByProviderID(ctxt context.Context, providerID string) (cloudprovider.Zone, error) {
```
Comment (bowei) [inline pkg/cloudprovider/providers/aws/aws.go]: I think the standard spelling is "ctx", not "ctxt"

**Comment by bowei** [comment]
If this change is only plumbing of the extra parameter, looks ok. There will be a follow on to actually plumb it through to the gce compute calls?

File: pkg/cloudprovider/providers/aws/aws.go (hunk 1 above)
Comment (cheftako) [reply pkg/cloudprovider/providers/aws/aws.go]: Your correct. Fixed.

**Comment by cheftako** [comment]
@bowei Your right this is just the plumbing. The plan was then to file issues to plumb it into every cloud providers calls.

File: pkg/cloudprovider/cloud.go [hunk 2]
```diff
@@ -56,9 +57,9 @@ type InformerUser interface {
 // Clusters is an abstract, pluggable interface for clusters of containers.
 type Clusters interface {
 	// ListClusters lists the names of the available clusters.
-	ListClusters() ([]string, error)
+	ListClusters(ctx context.Context) ([]string, error)
 	// Master gets back the address (either DNS name or IP address) of the master node for the cluster.
-	Master(clusterName string) (string, error)
+	Master(ctx context.Context, lusterName string) (string, error)
```
Comment (jhorwit2) [inline pkg/cloudprovider/cloud.go]: The c in clusterName became a casualty of the refactor :)

**Review by jhorwit2 — CHANGES_REQUESTED** [review]

File: pkg/controller/service/service_controller.go [hunk 3]
```diff
@@ -203,13 +207,15 @@ func (s *ServiceController) worker() {
 }
 
 func (s *ServiceController) processNextWorkItem() bool {
+	ctx, cancel := context.WithCancel(context.Background())
```
Comment (lavalamp) [inline pkg/controller/service/service_controller.go]: I am skimming this PR-- this looks questionable, should it be a TODO context, or re-use the one for the overall controller? Or maybe I misunderstand how contexts are supposed to be used?

File: pkg/controller/service/service_controller.go [hunk 4]
```diff
@@ -705,7 +711,7 @@ func loadBalancerIPsAreEqual(oldService, newService *v1.Service) bool {
 // syncService will sync the Service with the given key if it has had its expectations fulfilled,
 // meaning it did not expect to see any more of its pods created or deleted. This function is not meant to be
 // invoked concurrently with the same key.
-func (s *ServiceController) syncService(key string) error {
+func (s *ServiceController) syncService(ctx context.Context, key string) error {
```
Comment (lavalamp) [inline pkg/controller/service/service_controller.go]: Another context-free comment (pun intended), it feels weird to add this to every method, should it be a member variable instead?

**Comment by lavalamp** [comment]
I think the point of this is to fix the interface. What do you think about literally using context.TODO() at every call site? I think the specific context handling in various places deserves more consideration than we can give in a PR this size. And we can leave it up to the owners of the call sites to properly wire up context for their own code.

File: pkg/controller/service/service_controller.go (hunk 4 above)
Comment (cheftako) [reply pkg/controller/service/service_controller.go]: From the golang docs on context (https://golang.org/pkg/context/) "Do not store Contexts inside a struct type; instead, pass a Context explicitly to each function that needs it. The Context should be the first parameter, typically named ctx:"

File: pkg/cloudprovider/cloud.go (hunk 2 above)
Comment (cheftako) [reply pkg/cloudprovider/cloud.go]: Good catch. Fixed.

File: pkg/controller/service/service_controller.go (hunk 3 above)
Comment (cheftako) [reply pkg/controller/service/service_controller.go]: Controllers are an interesting variant. They are not requests from the outside, so at some point we need to create a root context. Given that I am creating a cancellable context, I did not want it spanning work items. I could create a background context at the worker level, pass it down and then make cancellable contexts from it at the processNextWorkItem() level.

**Comment by cheftako** [comment]
@lavalamp I think you may be right. I have been considering breaking this change into stages. However at a minimum I wanted to get a feeling for the scope of the change and talk to the effected teams. That way if there is a concern with the scope of the change we can deal with it up front rather than after having it half way rolled out. Happy to use this as a reference but make the actual pushes much smaller in scope. That having been said I would really like to get all of this + cloud provider implementations done in this release.

**Pull request closed without merging**